rfcs/20200804-configurable-filesystems.md

```diff
@@ -0,0 +1,135 @@
+| **RFC #**     | [NNN](https://github.com/tensorflow/community/pull/NNN) (update when you have community PR #) |
+| **Author(s)** | Sami Kama (kamsami@amazon.com)                                                                |
+| **Sponsor**   | Mihai Maruseac (mihaimaruseac@google.com)                                                     |
+| **Updated**   | 2020-08-04                                                                                    |
+
+## Objective
+
+The aim of this RFC to extend filesystem API to enable users to pass configuration parameters to tune the behavior of implementation to their use cases.
+
+## Motivation
+
+There are many FileSystem implementations in Tensorflow that enable interaction with various storage solutions. Most of these implementations have internal parameters that are suitable for generic use case but not necessarily optimal for all cases. For example accessing remote filesystems through multiple threads can improve the throughput if there is a high bandwidth connection to the remote thus increasing number of connections might be beneficial. On the other hand if the connection is slow, a higher number of threads will just waste resources and may even reduce the throughput. Depending on the resources available during the execution, users should be able to alter some of the parameters of the Filesystems to improve the performance of their execution. This can be especially useful for the the cases where the execution is data i/o bound.
+
+## User Benefit
+
+With this proposal users will be able to fine tune some parameters that developers expose through configuration API and get an improved perfomance for file i/o.
+
+## Design Proposal
+
+This proposal introduces two new methods to plugin api structure `TF_FilesystemOps` as shown below.
+
+```cpp
+struct TF_FilesystemOps{
+    // other members are ignored for brevity
+    void (*const get_filesystem_configuration)(char** serialized_config, int *serialized_length, TF_Status* status);
+    void (*const set_filesystem_configuration)(const char* serialized_config, int serialized_length, TF_Status* status);
+}
+```
+
+where `serialized_config` is a pointer to the buffer containing serialized human readable form of the protobuf object described below and `serialized_length` is the length of the buffer.
+
+For non-plugin based filesystems, FileSystem API can be extended similarly.
+
+```cpp
+class FileSystem{
+    public:
+    // existing methods are not shown.
+    Status GetConfiguration(std::unique_ptr<FilesystemConfig>* config);
+    Status SetConfiguration(std::unique_ptr<FilesystemConfig> new_configuration);
+}
+```
+
+Since each filesystem will likely to have different set of tunable parameters, a `FilesystemConfig` object can be used to unify the API and allow discovery of existing tunable parameters at runtime. We propose a protobuf object with the following schema
```

Comment:
I'd strongly prefer not to add protos to the API. Let's go with a plain struct, which is easier to make ABI safe?

Reply:
Hi Martin, Thanks for the first review. The ABI boundary will be crossed by prototbuf serialized to prototxt so it will be human readable plain text. It is to ensure that there is no ABI issue or problems due to small differences in protobuf library used in plugin or tensorflow. Plain struct will not be able to do easily unless we restrict the data types and arguments and how these data is serialized and deserialized in ABI compatible way. Does this alleviate some of your concerns about protobuf?

Reply:
How about

```cc
typedef struct key_val {
  char *key; // null terminated
  int version;
  int type_tag;
  union {
    int64 inv_val;
    double real_val;
    struct {
      char* buf;
      int buf_length;
    } buffer_val;
  } value; 
} key_val;
```

And then we can send an array of `key_val` elements.

It seems the option values can be either integer, reals or some buffers (always null terminated?). In that case, we can also have this API:

```cc
int64 get_int_option(const char* key);
double get_double_option(const char* key);
const char* get_char_option(const char* key);

void get_int_option(const char* key, int64 val);
void get_double_option(const char* key, double val);
void get_char_option(const char* key, const char* val);

int get_plugin_options(char** keys);
```

(modulo status codes). The second API proposal has the downside that user needs to first ask the plugin what options are available and then iterate over the returned array and set them. However, it is much more extensible (both API and ABI compatible by default -- already handled by the filesystem ABI/API compatibility layer). Whereas the struct-union variant from the beginning of this comment allows getting all the config options at once but requires adding a new version parameter.

Reply:
Actually, the second API allows setting options per path by adding a path argument to the methods

Reply:
Bools can be passed in via int options, lists can be passed via multiple calls or by adding 3 more API entries

Reply:
Hi @mihaimaruseac,

Thanks for  the comment. We could can also do that. But exposing these detail to user would make me uncomfortable since it will be too complicated for most users to use. We can use this (with addition of list methods) between layers and construct a class to wrap these in user facing (c/c++/python) side. That would be replicating part of protobuf functionality and I believe prototxt is more extensible and backward compatible. In case we change protobuf schema old plugins will still keep working but with this approach we will need recompilation of all plugins.

Reply:
We can have the function API proposed above between plugins and core TF (the C layer). Then, inside TF we can use a class/struct to hold them together and present the info to user in a class wrapper too (Python).

The C++/Python classes don't need to be backwards compatible from the point of view of the plugin, so we can change them as needed. But for the plugin interface, we need compatibility (both API and ABI).

The only people seeing the functional API above are plugin implementers and people who contribute to the filesystem layer in core TF.

Reply:
Using protos forces users to use protobuf, which has been a source of issues (it's not exactly a light dependency). Passing only serialized protos removes the ABI concerns, but using proto for configuration isn't something we love to do (maybe surprisingly, since it's all over our APIs, but that's just why we decided it might not be the best idea).

I agree that a set/get API is a bit of an overhead, but a struct would be fairly straightforward, I believe?

Reply:
@mihaimaruseac Yes I agree. Though your get_* api needs extension since having the name of the key doesn't tell you its type. With the struct approach, memory management might require some work but it is doable. I just believe some existing infrastructure like protobuf or alternatively json/yaml/xml serialized format would simplify things and reduce amount of code needed when developing plugins.

Reply:
@martinwicke I just believe using a library and human readable serialized data would be the best. It doesn't have to be protobuf. Only reason the protobuf suggested is that it is already used in TF. I am perfectly happy with other standard formats json/yaml/xml etc. Only advantage of protobuf over these is binary data and type information. Which could easily be implemented in other formats as well. 

I agree we can also use struct. But we need to address
- Memory ownership. Since there needs to be a query mechanism for existing values we will get the structs from plugins and need to manage the memory. There are methods in existing plugin API, although they are not documented in plugin RFC I believe, We can use these methods to manage memory, though we need to define lifetimes and ownership clearly.

- Multiple entries. We need to support passing a vector/list of values. It is possible to encode maps and other structures with lists.

- The documentation. We all know that developers hardly write documentation. Having a key and its value type is usually not sufficient to figure out what it means. We need to have a description field in the struct that documents the key.

- It would have been nice if everybody didn't have to implement Struct generation code from scratch but I can't see a way around it right now.

Reply:
We can use the following in the plugin interface

```cc
typedef struct TF_Filesystem_Option_Value {
  int type_tag;
  int num_values;
  union {
    int64 inv_val;
    double real_val;
    struct {
      char* buf;
      int buf_length;
    } buffer_val;
  } *values;  // owned
} TF_Filesystem_Option_Value;

typedef struct TF_Filesystem_Option {
  char* name; // null terminated, owned
  char* description; // null terminated, owned
  int per_file;  // bool actually, but bool is not a C type
  TF_Filesystem_Option_Value *value;  // owned
} TF_Filesystem_Option;
```

This handles both documentation (provided by plugin) and list values (I'm not conviced we need maps here, but it could be possible to extent the `union` to add them). Plus, since above is in filesystem interface, it is used by core TF and we can provide functions in the filesystem interface to manage these structs. This way, plugins only need to know the layout of the structs and read/write to them directly.

Memory allocation will have to come from the plugin's side, using the same routines plugins currently use for the other filesystem operations. It's the only way this can work on windows so we cannot choose anything else.

We probably need a new ABI number but adding it to the filesystem registration function would be an ABI breakage. We can handle this by increasing existing ABI numbers from 0 to 1 or we can use the options ABI number only in the function that core TF uses to get filesystem's default options.

Plugins will need to fill these structs in with all the options they support, change them (if plugin allows changing them), and read from them when implementing operations that depend on these options. Core TF only needs to propagate these options up to the C++/Python layers and display them to users.

I would expect that a user trying a new filesystem plugin will do something like

```python
options = tf.io.get_filesystem_options_for_scheme("new_uri_scheme://")
tf.io.display_filesystem_options(options)  # will print current options and the help text too
tf.io.set_option(options, key, value)
...
```

We have a lot of pointer chasing in this structure but since this is IO I doubt it will result in significant performance degradation.